### PR TITLE
incompatible_rank_assignment

### PR DESCRIPTION
**Pull request by AnshMehta1**

### CODE REVIEW

File: tests/tests.toml [hunk 1]
```diff
@@ -4168,6 +4168,6 @@ asr = true
 filename = "implied_do_loop1.f90"
 run = true
 
-[[test]]
```
Comment (kmr-srbh) [inline tests/tests.toml:4171]: Please include this line.

File: tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.stderr [hunk 2]
```diff
@@ -1,9 +1 @@
-code generation error: asr_to_llvm: module failed verification. Error:
```
Comment (kmr-srbh) [inline tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.stderr:1]: Please remove this file. This reference file is causing some issues. I will try to update it in a different PR.

File: tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.json [hunk 3]
```diff
@@ -5,9 +5,9 @@
     "infile_hash": "3e65758e10744eea8d101be6437af3724c0ec32c832e29da968fc22d",
```
Comment (kmr-srbh) [inline tests/reference/run_dbg-runtime_stacktrace_01-dcc746e.json:5]: Please remove this file from the PR.

**Review by kmr-srbh — COMMENTED** [review]
Left some comments. Let's get the CI to pass, I will review closely later.

**Comment by AnshMehta1** [comment]
Made the changes , but still my ci tests are not passing.

File: tests/tests.toml [hunk 4]
```diff
@@ -4171,3 +4171,9 @@ run = true
 [[test]]
 filename = "multi_error1.f90"
 ast = true
+
+[[test]]
+filename = "./errors/incompatible_rank.f90"
+asr = true
+
```
Comment (kmr-srbh) [inline tests/tests.toml:4178]: Let's add the `run = true` field to execute the code and obtain the error message.

**Comment by kmr-srbh** [comment]
> Made the changes , but still my ci tests are not passing.

Great! Now let's fix what broke the CI. The output says:

``
compiler_tester.tester.RunException: Testing with reference output failed.
../integration_tests/arrays_reshape_14.f90 * asr
The JSON metadata differs against reference results
Reference JSON: tests/reference/asr-arrays_reshape_14-913f34e.json
Output JSON:    tests/output/asr-arrays_reshape_14-913f34e.json
Omitting 6 identical items
```

This means that the output with changes in this PR, do not match the output with `main`. I had a look into it and it seems we caught a different bug. I reported it at #4798.

Could you please confirm if the code given in #4798 throws an error with your changes?

File: src/lfortran/semantics/ast_common_visitor.h [hunk 5]
```diff
@@ -2930,6 +2930,17 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                    int LHS_rank = ASRUtils::extract_n_dims_from_ttype(type);
+
+                    ASR::ttype_t* value_expr_type = ASRUtils::expr_type(value);
+                    int RHS_rank = ASRUtils::extract_n_dims_from_ttype(value_expr_type);
+                    
+                    if(LHS_rank != RHS_rank){
+                        throw SemanticError("Incompatible ranks "+
+                            std::to_string(LHS_rank) + " and " +
+                            std::to_string(RHS_rank)+" in assignment", x.base.base.loc);
+                    }
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:2942]:
```fortran
program main
integer :: x(2, 2) = reshape([1,2,3,4], [2,2])
end program
```

This shall work but will throw an error with your change or something similar.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 6]
```diff
@@ -2930,6 +2930,17 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     init_expr = ASRUtils::EXPR(tmp);
                     value = ASRUtils::expr_value(init_expr);
 
+                    int LHS_rank = ASRUtils::extract_n_dims_from_ttype(type);
+
+                    ASR::ttype_t* value_expr_type = ASRUtils::expr_type(value);
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:2935]:
```suggestion
                    ASR::ttype_t* value_expr_type = value ? ASRUtils::expr_type(value): value;
```

**Pull request closed without merging**